Car_Localizer/Car_Localizer/Training/custom_functions.py

```diff
@@ -0,0 +1,49 @@
+import os
+os.environ['KMP_DUPLICATE_LIB_OK'] = 'True'
+import torch
+
+#Initialize function to calculate the mean and std - used in normalization process:
+def get_mean_and_std(dataloader):
+    """Returns a mean and std of train dataset in order to perform normalization on final train dataset.
+            Parameters
+            ----------
+            dataloader : torch tensor collections
+                train_raw dataset transformed to torch tensor and prev-resized
+            Returns
+            -------
+                Mean and Std in  list format
+    """
+    channels_sum, channels_squared_sum, num_batches = 0, 0, 0
+    for data, _ in dataloader:
+        # Mean over batch, height and width, but not over the channels
+        channels_sum += torch.mean(data, dim=[0, 2, 3])
+        channels_squared_sum += torch.mean(data ** 2, dim=[0, 2, 3])
+        num_batches += 1
+    # mean: the following function
+    mean = channels_sum / num_batches
+    # std = sqrt(E[X^2] - (E[X])^2)
+    std = (channels_squared_sum / num_batches - mean ** 2) ** 0.5
+    mean = torch.Tensor(mean).tolist()
+    std = torch.Tensor(std).tolist()
+    return mean, std
```

Comment:
we should keep everything in PyTorch's tensor manner
so just return the tensor instead a list

```
'''
  Data normalization
  https://www.baeldung.com/cs/normalizing-inputs-artificial-neural-network
'''
def get_mean_std(data:torch.tensor) -> torch.tensor: #get_mean_std 
    assert torch.is_tensor(data), "Example message here..."
    return data.mean(), data.std()
#or
def data_normalization(data:torch.tensor, mean_zero:bool=False) -> torch.tensor: #z_score 
    mu, sigma = data.mean(), data.std()
    return (data - 0.) / sigma if mean_zero else (data - mu) / sigma
```
you can simply use PyTorch's function based

Reply:
I cannot see any other potential processing way how to resolve it properly.
This time I did not want to use custom functions to accomplish it.
#------------------------------------------------------------------------#
Check it pls:
train_loader_raw = DataLoader(dataset=train_dataset_raw, shuffle=False, batch_size=len(train_dataset_raw))
data_raw = next(iter(train_loader_raw))

and after that, data_raw[0].mean() and data_raw[0].std()

Just to clarify:
train_dataset_raw is required to load the inputs and perform the transformations methods like Resize(64,64) as also ToTensor().

Reply:
ok, got it